lib/node_starter/queue_subscribe.rb

```diff
@@ -64,21 +54,24 @@ def run(delivery_info, payload)
         params['enqueue_data'],
         node_api_address(uuid))
 
-      begin
-        starter.start_node_process
-        @consumer.ack(delivery_info)
-        @shutdown_consumer.register_node(params['build_id'])
-      rescue => e
-        NodeStarter.logger.error "Node #{params['build_id']} spawn failed: #{e}"
-        @consumer.reject(delivery_info, true)
-        return
-      end
+      @reporting_publisher.receive(params['build_id'])
+      starter.start_node_process
+      @consumer.ack(delivery_info)
+      @shutdown_consumer.register_node(params['build_id'])
+      @reporting_publisher.start(params['build_id'])
+      wait_for_node(starter, params['build_id'])
+    rescue => e
```

Comment:
I'd suggest to divide this block to two:
1. until the `start_node_process` we can reject the amqp message.
2. the node process is running - from this point we htave to set this build as `errored` on any error message.

btw, the first block is so easy to handle errors: we have to send `reset` event to the travis-hub at least; and set to error on last try. For make things simple I suggest to set errord as well. e.g. we can use one block which settings `error` state on exceptions.